### ISSUE TITLE
Tracking log Events damaging site performance

### ISSUE THREAD

**Issue by luancschmitz** [Smile-SA/elasticsuite#3054]
I received a problem from Adobe support related to tracking_log indexes.

It seems to me that they are consuming a lot of the store.

Below is what the Adobe team told me:

`$ php bin/magento config:show catalog/search/engine
elasticsuite
`

`$ curl -s localhost:9200/_cat/indices | wc -l
141
`

`$ curl -s localhost:9200/cluster/health?pretty | grep -i "active*shard"
"activeshards" : 1269,
"activeshardspercentas_number" : 100.0
r25dxhzx4nzum@i-bb24c8159eef891d:~$ curl -sS "localhost:9200/_cat/nodes?h=heap*&v"
heap.current heap.percent heap.max
2.1gb   21    9.9gb
6.3gb   63    9.9gb
6.7gb   67    9.9gb`

`r25dxhzx4nzum@i-bb24c8159eef891d:~$ curl -s localhost:9200/_cat/indices | grep -i tracking
green open magento2_store_view_rn_tracking_log_event_202309        GX8AKq_iSEakbtKmTKuDJQ 3 2  1936   0    4.7mb   1.7mb
green open magento2_store_view_es_tracking_log_event_202309        ms5PtF_XRBuA4O3h_P5b5w 3 2  4157   0    7.6mb   2.4mb
green open magento2_store_view_mt_tracking_log_session_202309      JoD0C_XgTrK45I3J7tYqPA 3 2    24   0  352.9kb 117.6kb
green open magento2_store_view_pe_tracking_log_session_202309      rOA7h7PmRWqrjRlkpfp11Q 3 2     4   0   60.3kb  20.1kb
green open magento2_store_view_al_tracking_log_session_202309      IFi61MxUT8qoqVl3s_AyLg 3 2     2   0   72.8kb  10.8kb
green open magento2_store_view_am_tracking_log_event_202309        eJJqvgFATPqzs8xHPMDzug 3 2 13225   0   17.9mb   5.8mb
green open magento2_store_view_pa_tracking_log_session_202309      UNyLZFMgSdCvas1ChYcDtA 3 2   977  79    1.4mb 472.5kb
green open magento2_store_view_ba_tracking_log_session_202309      imVL2JgiTu-jr0GKsx9Jqw 3 2   918  38    1.3mb   445kb
green open magento2_store_view_pa_tracking_log_event_202309        jCB68CD1T1inTwJEDUNHVg 3 2  9647   0   16.8mb   6.1mb
green open magento2_store_view_es_tracking_log_session_202309      XTGyryyESmS5kP1q1ZDyXA 3 2   726  31    1.2mb 426.4kb
green open magento2_store_view_rj_tracking_log_session_202309      5Fpn0C8tT4eBT7BXKCnngQ 3 2    39  10  387.2kb 130.8kb
green open magento2_store_view_sc_tracking_log_event_202309        S7sWzlntQ8e3JQuTPiJN3w 3 2     6   0    237kb    79kb
green open magento2_store_view_ro_tracking_log_session_202309      4ZcvtGLfTmit4tlUo149dQ 3 2   538  33 1017.9kb 339.3kb
green open magento2_store_view_rr_tracking_log_event_202309        D43C59stQE-N2fuU8zi9qQ 3 2  2351   0      6mb   1.8mb
green open magento2_store_view_ce_tracking_log_event_202309        cBQbSpLkRxyrowq5koarog 3 2  5505   0   10.1mb   3.3mb
green open magento2_store_view_se_tracking_log_session_202309      Uz5pPWkLTxuRXpozs6KU0w 3 2     5   0   74.8kb  24.9kb
green open magento2_store_view_am_tracking_log_session_202309      bo-jVKN3R9axXrjVHMfSdw 3 2  1687  76    2.2mb 753.5kb
green open magento2_store_view_pi_tracking_log_event_202309        2PEow9e7SHKVBLS2H9VhFg 3 2    52   0    735kb   245kb
green open magento2_store_view_mg_tracking_log_event_202309        CgrNW54mTZu6KXlsGtTGgg 3 2    20   0  739.6kb 239.9kb
green open magento2_default_tracking_log_session_202309            mOdlW541Q7qIA_lNvagX6w 3 2   257   6  600.8kb 206.8kb
green open magento2_store_view_se_tracking_log_event_202309        laaU0eZESiSa_SWpcy4CkA 3 2     5   0  199.8kb  66.6kb
green open magento2_store_view_mt_tracking_log_event_202309        -SmaOh_bQS6Ri0oTFaGxsA 3 2    76   0      1mb 335.5kb
green open magento2_store_view_ms_tracking_log_event_202309        9GZLIa9HR6-yxbj1khbx7w 3 2    11   0  428.5kb 142.8kb
green open magento2_store_view_sp_tracking_log_session_202309      0H5ATo5fQf6IEbVcAZIo8g 3 2     5   0  150.3kb  50.1kb
green open magento2_store_view_ma_tracking_log_session_202309      JIWrZfy9Tpq5q84LfOWRUw 3 2  2726  85      3mb     1mb
green open magento2_store_view_to_tracking_log_event_202309        QeXUcArQQiqc4LAzvgR7zw 3 2     4   0  129.1kb    43kb
green open magento2_store_view_rn_tracking_log_session_202309      2-5ImJpSSNmhXAyRGEo48A 3 2   511  11  856.4kb 289.6kb
green open magento2_store_view_ac_tracking_log_session_202309      imNDMJsXSM6Kpywy9WbAHA 3 2   203  28    648kb   216kb
green open magento2_store_view_mg_tracking_log_session_202309      CC3J6krKTHGhff0PGl2lSw 3 2     4   0   96.7kb    21kb
green open magento2_store_view_ma_tracking_log_event_202309        FmrEQWJ8Q9-3Xqhe4fpJTg 3 2 10534   0   17.1mb     5mb
green open magento2_store_view_sp_tracking_log_event_202309        HDwcRMRxQa-O_xrMIE4_HQ 3 2    15   0  568.9kb 189.6kb
green open magento2_store_view_ro_tracking_log_event_202309        5sCxUlpdQ563t5a-Ka8lZw 3 2  3236   0    6.8mb   2.2mb
green open magento2_store_view_ba_tracking_log_event_202309        BPSEyVmqRlWa5HGT5TCQNQ 3 2  4974   0    8.1mb   2.7mb
green open magento2_store_view_ce_tracking_log_session_202309      Bdinm3KMRZmZGFvmMwrVhw 3 2   671  34    1.1mb 396.8kb
green open magento2_store_view_sc_tracking_log_session_202309      0iNiQGnhRmWPphjsQ2tORg 3 2     6   0   88.4kb  29.4kb
green open magento2_store_view_pr_tracking_log_session_202309      VuFbSKgRQzymawRXMJcyQA 3 2     5   0  100.4kb  33.4kb
green open magento2_store_view_al_tracking_log_event_202309        Y4w_FoXuTuu5A1UTau-roQ 3 2     7   0  279.5kb  93.1kb
green open magento2_store_view_rr_tracking_log_session_202309      eaFbSa1CR96XRLAot9oo3A 3 2   660  14      1mb 372.8kb
green open magento2_store_view_rj_tracking_log_event_202309        rR5BLhaYSIqlbUvkkSNxuA 3 2   308   0    1.5mb 453.6kb
green open magento2_store_view_df_tracking_log_session_202309      cgT822CaR4WzqVxGiM-xVQ 3 2     1   0  181.8kb  62.8kb
green open magento2_store_view_df_tracking_log_event_202309        M4ViEHToQJm0P0TP6nU8HQ 3 2    19   0  757.8kb 252.6kb
green open magento2_store_view_ac_tracking_log_event_202309        DBlk4BPMTJm_M0MfiH8mmA 3 2  1069   0    3.9mb   1.3mb
green open magento2_store_view_pi_tracking_log_session_202309      WueZjMekTkuHtdSvervNAg 3 2    13   0  189.2kb    63kb
green open magento2_store_view_ms_tracking_log_session_202309      ZEAu5_ELQo2uLHD2MXn37A 3 2     6   0   89.2kb  29.7kb
green open magento2_store_view_ap_tracking_log_session_202309      -xCaEXz3QqKZRlGLaMbNkA 3 2   668  21  988.5kb 329.5kb
green open magento2_store_view_to_tracking_log_session_202309      aw9ZjZxPSA2k1BM1GDhOPA 3 2     2   0   31.5kb  10.5kb
green open magento2_store_view_pr_tracking_log_event_202309        bHTL-NGaSjeuI7-K3Foznw 3 2     9   0  296.6kb  98.8kb
green open magento2_default_tracking_log_event_202309              c6s6x4-gSbyUZT34mcThXQ 3 2   383   0    1.7mb 500.6kb
green open magento2_store_view_pe_tracking_log_event_202309        z17IRQaESc2N8BcX3sRWbw 3 2     4   0    160kb  53.3kb
green open magento2_store_view_ap_tracking_log_event_202309        PMJGilFlQ9Oz4aa5N-5RKQ 3 2  3968   0      7mb   2.3mb
`

Our website currently has 28 stores.

Magento Version: 2.4.3-p3
Smile version: 2.11.3.1

For now, I have disabled this functionality to see if the site no longer experiences this problem.

But I would like to know what interferes with this function

**Comment by luancschmitz** [comment]
I was using the default value. I ended up changing it as shown in the image below:

![Captura de Tela 2023-09-25 às 12 06 06](https://github.com/Smile-SA/elasticsuite/assets/13907586/8c5fc6ad-4cf3-46fa-b1b8-3112e467aa85)

But the problem persisted.

Sorry for the question, how can I check the size?

I used the command:
curl -XGET localhost:9200/_stats

but I couldn't find the correct size

**Comment by vahonc** [comment]
@luancschmitz 

Also, you should see a warning message at the top of the page, if your configuration is not correct.

![Screenshot from 2023-09-25 18-37-28](https://github.com/Smile-SA/elasticsuite/assets/67791361/53a8b36d-8b76-435b-8ae6-bc594dcdfa03)

BR,
Vadym

**Comment by luancschmitz** [comment]
here is the result

![Captura de Tela 2023-09-25 às 13 41 51](https://github.com/Smile-SA/elasticsuite/assets/13907586/9af6bedd-36d9-449c-bbd5-9d9fb671873f)

**Comment by luancschmitz** [comment]
@vahonc 

Hello,

Looking at the elasticsearch message the following message is displayed:

The number of shards configured for Elasticsuite is incorrect. You do not need to use 3 shards since your biggest Elasticsuite index is only 9 MB.
Click here to go to the Elasticsuite Config page and change your Number of Shards per Index parameter according to our wiki page.


But I ended up leaving the 3 shads on Adobe's recommendation. What was probably causing the elasticsearch headsize was the tracking_log indexes, but I ended up disabling them.

**Comment by rbayet** [comment]
Hello @luancschmitz,

Adobe recommendations are set into stone, they never take into consideration the actual need in terms of performance and/or accuracy versus the actual size of the index.
Yes, on a Adobe OpenSearch/Elasticsearch cluster, you should have 2 replicas but you almost never need 3 primary shards.

The pro : 
* supposedly helps distributes the load between the three master/data nodes
  * which is useful to Adobe because non-production cluster are often *shared* between your integration/staging instances

The cons : 
* the master has to wait anyway the answer from the other shards when doing a query, especially when you want the exact number of documents
* Elasticsuite has to send as many "term vector/pre-analysis requests" as there primary shards
* you inflate dramatically the number of shards when you have a lot of stores

Granted, Elasticsuite *should* provide a way to configure different number of primary shards depending on the data you index (products, categories, thesaurus, behavioral data), because that number of primary shards applies to all, even IF and WHEN you need multiple primary shards, it is usually only for the catalog_product indices which contains the most data.

So here's what happened on your instance considering you initially where configured with primary shards=1 and replicas=2.
Per store : 
* products : 1 * (2+1) shards = 3 shards
* categories : 1 * (2+1) shards = 3 shards
* thesaurus : 1 * (2+1) shards = 3 shards
* behavioral data (let's say you kept 12 months retention) : maximum of 12 months * 2 indices (tracking log, tracking session) * 1 * (2+1) shards = 72 shards
Total per store: 81 shards 
Total per store per node : 27 shards

**With 27 shards per node, you would have been in very reasonnable grounds with 10GB of heap (memory)** allocated to each node (the rule of thumb is "you need at least 1GB per 20 shards per node just for the shards overhead") **if you had only one store**

Now, if I multiplied by 28 stores, that's (27 * 28) 756 shards per node.
*You would need at least 38 GB of heap per node **just** for the shards overhead.*

But Adobe Cloud applied/convinced you to change the number of primary replicas to 3...
Per store : 
* products : 3 * (2+1) shards = 9 shards
* categories : 3 * (2+1) shards = 9 shards
* thesaurus : 3 * (2+1) shards = 9 shards
* behavioral data (let's say you kept 12 months retention) : maximum of 12 months * 2 indices (tracking log, tracking session) * 3 * (2+1) shards = 216 shards
Total per store: 243 shards 
Total per store per node : 81 shards

**Multiply 81 shards per 28 stores, that's 2268 shards per node ;(**
I guess if you ask Adobe for (2268 / 20) 113 extra GB of Heap per node, they'll choke and if they don't, you will when presented with the bill. :D

Long story short : 
* considering the size of your catalog_product indices, you do not need 3 primary shards, get back to 1 and reindex everything
* if you want to preserve memory resources you can either
  * disable alltogether the tracking of Elasticsuite
  * or reduce the "retention delay" in months (2 indices per month per store)
     * if you're not planning in exploiting the behavioral data and are just interested in catching 0 result queries or mispelled queries and you frequently have a look at Elasticsuite analytics dashboard, even 1 month of retention delay could be beneficial
  * disabling the Elasticsuite tracker or reducing the retention delay, say from 12 to 2 months will not necessarily automatically remove your behavioral data indices, so you might have to delete them manually to free memory

Deleting indices on Elasticsearch/OpenSearch cluster (use that command with care) : 
*To delete ONE tracking log event index* : 
`curl -XDELETE localhost:9200/magento2_store_view_ac_tracking_log_event_202309`

*To delete ONE tracking log session index* :
`curl -XDELETE localhost:9200/magento2_store_view_ac_tracking_log_session_202309`

*To delete ALL tracking indices of your "store_view_ac" store* : 
`curl -XDELETE localhost:9200/magento2_store_view_ac_tracking_*`

*To delete ALL tracking indices of that Magento instance* : 
`curl -XDELETE localhost:9200/magento2_*_tracking_*`

Regards,

**Comment by luancschmitz** [comment]
Hello @rbayet , thank you very much for the response. Now I understand it well. I ended up leaving tracking_log disabled, I had already deleted all tracking log indexes for all stores, and returned with the primary shard setting to 1.

so from what I understand, do you advise increasing the heapsize to 38 gb?

print of how you are today:

![Captura de Tela 2023-09-25 às 18 48 34](https://github.com/Smile-SA/elasticsuite/assets/13907586/0e5c87f7-6ab4-40d6-a6b5-1be426da3ef9)

**Comment by rbayet** [comment]
Hello @luancschmitz,

The 38 GB of heap would have been if you intended to keep behavioral data enabled for all stores and with a 12 months retention delay.

With the tracking totally disabled and the tracking indices removed, and with primary shard set to 1, you're back to 
Per store :
* products : 1 * (2+1) shards = 3 shards
* categories : 1 * (2+1) shards = 3 shards
* thesaurus : 1 * (2+1) shards = 3 shards
Total per store: 9 shards
Total per store per node: 3 shards
Total all stores per node: 3*28 = 84 shards

84 shards per node divided by 20 gives 4.2 GB of heap per node for shards overhead.
You have 10GB of heap per node, so it's looks ok as it is.

If you decided to re-enable the tracker with 1 month of retention delay, it would add
Per store: 2 indices (tracking_log, tracking_session) * 1 month * (2+1) shards = 6 shards
Per store per node: 2 shards
Total all stores per node: 2*28 = 56 shards

That would make a total of 84 + 56 = 140 shards per node, so you would need just 7 GB of heap *just* for the shards overhead.
7GB of 10GB of heap, we are in the "Not great, not terrible" area.

Whatever you decided, depending on your traffic, you might want to look from time to time at https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-nodes-stats.html to inspect "jvm.gc.collectors[young,old]".
Basically, if jvm.gc.collectors.old.collection_count and/or jvm.gc.collectors.old.collection_time_in_millis skyrockets this means a node is more and more frequently doing deep garbage collections and/or taking more and more time doing it.

In a perfect world, that's what a company providing Elasticsearch/OpenSearch hosting would do besides just reacting when a node crashes because he reached its heap size limit.

Regards,
